Source/WebCore/accessibility/AccessibilityNodeObject.cpp

```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+    if (auto* image = dynamicDowncast<HTMLImageElement>(*node); image && objectCache) {
+        // https://github.com/w3c/aria/pull/2224
+        // Per html-aam, <img> elements that are unlabeled (e.g., alt attribute, ARIA, title) derive accname
+        // from an ancestor figure's <figcaption> if and only if the <figure> does not contain other flow content (besides the <figcaption>).
+        const AtomString& alt = image->attributeWithoutSynchronization(altAttr);
+        const AtomString& title = image->attributeWithoutSynchronization(titleAttr);
+
+        if (alt.isEmpty() && title.isEmpty()) {
+            for (auto* ancestor = this->node()->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
+                if (auto* element = dynamicDowncast<Element>(ancestor); element && element->hasTagName(figureTag)) {
+                    bool figureHasFlowContent = false;
+                    // Iterate over the direct children of the <img>'s ancestor <figure> for any common
+                    // flow content, including non-whitespace text nodes.
+                    for (auto* node = ancestor->firstChild(); node; node = node->nextSibling()) {
+                        if (auto* element = dynamicDowncast<Element>(node); element && !element->hasTagName(figcaptionTag) && !element->hasTagName(imgTag)) {
+                            figureHasFlowContent = element->hasTagName(preTag)
+                                || element->hasTagName(canvasTag)
+                                || element->hasTagName(codeTag)
+                                || element->hasTagName(blockquoteTag)
+                                || element->hasTagName(divTag)
+                                || element->hasTagName(olTag)
+                                || element->hasTagName(pictureTag)
+                                || element->hasTagName(pTag)
+                                || element->hasTagName(spanTag)
+                                || element->hasTagName(ulTag);
+                        }
+
+                        if (auto* text = dynamicDowncast<Text>(node); text && !text->data().containsOnly<isASCIIWhitespace>())
+                            figureHasFlowContent = true;
+                    }
+                    // If the <figcaption> is non-empty and no flow content is present in the <figure>,
+                    // the <img> derives accname from the <figcaption>.
+                    if (auto* figureObject = objectCache->getOrCreate(ancestor); figureObject && figureObject->isFigureElement()) {
+                        auto* figcaptionAXNodeObject = dynamicDowncast<AccessibilityNodeObject>(*figureObject)->captionForFigure();
+                        if (figcaptionAXNodeObject && !figureHasFlowContent && !figcaptionAXNodeObject->isHidden())
+                            textOrder.append(AccessibilityText(accessibleNameForNode(*figcaptionAXNodeObject->node()), AccessibilityTextSource::Alternative));
```

Comment:
Couple thoughts here:

1. The only time this block does something is when `!figureHasFlowContent`. We should check that first before doing `AXObjectCache::getOrCreate`
2. figcaptionAXNodeObject is a bit of a long name. I think since it's used in such a small block, calling it just `caption` is fine.
3. In this case, because you don't check the result of the `dynamicDowncast`, you may as well use `downcast`. The difference is that the latter will cause a crash if the type isn't what is expected, whereas `dynamicDowncast` just returns `nullptr`. But you're not null-checking the result anyways, which is probably fine (we know this object is associated with a figure element, which must be an `AccessibilityNodeObject`).